### PR TITLE
Add multi-tenancy support for signals in the engine

### LINKED ISSUE #13336: Zeebe supports multi-tenancy for BPMN signal events
### Description

This issue covers multi-tenancy support for signal broadcasting in Zeebe. 

```[tasklist]
### Task breakdown
- [ ] https://github.com/camunda/zeebe/issues/14277
- [ ] https://github.com/camunda/zeebe/issues/13558
- [x] The `SignalSubscriptionState` can perist the `tenantId`
- [x] Engine can process `SignalRecord` commands with tenant id
- [x] Multi-tenancy data migration is provided for ProcessingStates used by signal broadcasting
- [ ] https://github.com/camunda/camunda-docs/issues/2820
- [x] ElasticSearch/OpenSearch signal templates include tenantId
- [ ] Notify the Operate team that multi-tenancy is supported for signal broadcast in Zeebe
```

### Blocked by
- https://github.com/camunda/zeebe/issues/13238

### PR DESCRIPTION
**Pull request by koevskinikola**
## Description

Add multi-tenancy support for signals in the engine

## Related issues

closes #13336 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [x] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [x] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [x] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Thanks @koevskinikola! 🚀 

I have quite a few comments, mostly regarding some issues in the state and migration. Please have a look at those and let me know if you have questions.

File: engine/src/main/java/io/camunda/zeebe/engine/state/signal/DbSignalSubscriptionState.java [hunk 1]
```diff
@@ -96,23 +114,24 @@ public void visitBySignalName(
...
+    subscriptionKey.wrapLong(elementInstanceKey);
     visitSubscriptions(visitor);
   }
 
   private void visitSubscriptions(final SignalSubscriptionVisitor visitor) {
     subscriptionKeyAndSignalNameColumnFamily.whileEqualPrefix(
-        this.subscriptionKey,
+        subscriptionKey,
         (key, value) -> {
+          // the signalName and tenantIdKey are wrapped on look-up in the background
```
[the remark this answers is not in the document]
Comment (koevskinikola) [reply engine/src/main/java/io/camunda/zeebe/engine/state/signal/DbSignalSubscriptionState.java:132]: Sure, it definitely makes things more readable.

**Comment by koevskinikola** [comment]
@remcowesterhoud thanks for all the review feedback. I've implemented all of it. Have a look when you have the time.

**Review by remcowesterhoud — APPROVED** [review]
LGTM! 

bors merge

**Comment by koevskinikola** [comment]
bors retry

**Pull request merged**